02-intermediate-ruby/api-post.md

```diff
@@ -0,0 +1,290 @@
+  BASE_URL = 'https://slack.com/api/'
+
+  # Code goes here
+
+end
+```
+
+In this exercise we will build some tests for a `send_msg` method in the `SlackApiWrapper` class and then build out the functionality to satisfy the tests.
+
+### Our First Test - Nominal Case
+
+For our first test we will build a test case for sending a message to Slack successfully.  We will need to set up the testing first.  Add the following to `specs/spec_helper.rb`
+
+```ruby
+# spec_helper.rb
+require 'dotenv'
+Dotenv.load
+
+require 'minitest/autorun'
+require 'minitest/reporters'
+require 'vcr'
+require 'webmock/minitest'
+require_relative '../lib/slack_api_wrapper'
+
+Minitest::Reporters.use! Minitest::Reporters::SpecReporter.new
+
+VCR.configure do |config|
+  config.cassette_library_dir = 'specs/cassettes' # folder where casettes will be located
+  config.hook_into :webmock # tie into this other tool called webmock
+  config.default_cassette_options = {
+    record: :new_episodes,    # record new data when we don't have it yet
+    match_requests_on: [:method, :uri, :body] # The http method, URI and body of a request all need to match
+  }
+  # Don't leave our Slack token lying around in a cassette file.
+  config.filter_sensitive_data("SLACK_TOKEN") do
+    ENV["SLACK_TOKEN"]
+  end
+end
+```
+
+This code configures VCR just like we did when we tested `GET` requests.  We have added `config.filter_sensitive_data("SLACK_TOKEN")` so that anytime our VCR sees our slack token it will replace it with the words "SLACK_TOKEN".  **This will prevent the token from appearing in git**.
+
+We can add our Slack token to a `.env` file in the project's root folder like this:
+
+```
+SLACK_TOKEN=<TOKEN_GOES_HERE>
+```
+
+![.env file picture](images/dotenv.png)
+
+Next we can write our first test using VCR.
+
+```ruby
+# slack_api_wrapper_spac.rb
+require_relative 'spec_helper'
+
+describe SlackApiWrapper do
+  it "can send a valid message" do
+    VCR.use_cassette("slack-posts") do
+      response = SlackApiWrapper.send_msg("Hey I can post messages!", "YOUR-CHANNEL-NAME")
+      expect(response).must_equal true
+    end
+  end
+end
+```
+
+This test calls a class-method called `send_msg` and sends it a message and channel name.  When we run our tests we get one test error because the method does not yet exist.  Now we can write `send_msg` to get the test to pass.
+
+With `HTTParty`, you can use the `post` method to send `POST` requests.  The `post` method takes two arguments.
+
+1. The URL of the post request, like `https://slack.com/api/chat.postMessage`.
+1. A hash with the following key-value pairs.
+    - `headers` a hash of key-value pairs to put in the request header.
+    - `body` a hash of key-value pairs for the data sent to the API.
+
+Like the `get` method, `post` returns an HTTP response from the API.
+
+For example:
+
+```ruby
+response = HTTParty.post("http://adapets.org/pets", {
+  headers: { 'Content-Type' => 'application/json' },
+  body: {
```

Comment:
Blugh, Chris definitely had it with the syntax without the `{ ... }` but I recommended to add it because I don't know if we talk about that syntax by this point. Sorry!

Reply:
Oof, sorry to tug you around. I guess either way would be fine.

I would still recommend removing the `headers` argument.